[Merged by Bors] - Fix memory leaks in p2p

lrettig opened the pull request:
## Motivation
Closes #2313

## Changes
- Correctly close msgConn in UDP listener when closing UDP connection wrapper
- Fix P2P incoming connection handler to correctly handle max conns

## Test Plan
Adds regression tests for both fixed memory leaks:
- test to make sure that new network connections aren't added beyond MaxPendingConnections, until one is released
- test to make sure that msgConnection is closed when a UDP connection is evicted

## TODO
<!-- This section should be removed when all items are complete -->
- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [ ] Update documentation as needed

## DevOps Notes
<!-- Please uncheck these items as applicable to make DevOps aware of changes that may affect releases -->
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [ ] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

lrettig committed:
Fix memory leaks in p2p

lrettig committed:
Linter

y0sher commented on p2p/net/network.go:
```diff
@@ -304,16 +307,21 @@ func (n *Net) accept(listen net.Listener) {
-			go c.beginEventProcessing()
+
+			// Asynchronously handle the new connection
+			func() {
```
I really don't understand, you meant to run this in a goroutine?

lrettig replied:
Yes. Otherwise, the `defer` runs immediately when the `func` ends, which happens immediately, since `go c.beginEventProcessing()` is async. That's not what we want. We want to defer the `pending <- struct{}{}` until `c.beginEventProcessing()` returns.

lrettig replied:
Yes, typo, will fix! Thanks

lrettig committed:
Fix typo

Spotter: @y0sher

lrettig commented:
Marking as draft for now as I still need to add regression tests

y0sher commented on p2p/net/network.go:
```diff
@@ -304,16 +307,21 @@ func (n *Net) accept(listen net.Listener) {
+			// Asynchronously handle the new connection
+			go func() {
+				// Return the new connection token once the connection is closed
+				defer func() { pending <- struct{}{} }()
```
This was at the first line in order to catch the case where `setupIncoming` failed, did you change it purposely?

lrettig replied:
It has to live in the same goroutine as `beginEventProcessing()` - that's the source of the memory leak, that it would run before `beginEventProcessing()` finished. But, you're right, I will separately handle the case where `setupIncoming` failed, good catch.

lrettig committed:
Bugfix to pass tests

msgcon does not need its own closer. It's now closed from inside the UDP
wrapper. Otherwise, we get a panic on trying to close a closed channel.

lrettig committed:
Test that msgConn is closed on UDP conn eviction

Mocks the msgConn in the test. Added a very simple interface to make
this easier.

lrettig committed:
Another test fix

Need to init this channel manually in test

noamnelke commented on p2p/net/msgcon.go:
```diff
@@ -17,17 +17,18 @@ import (
+	r          io.Reader
+	wmtx       sync.Mutex
+	w          io.Writer
+	//closer      io.Closer
```
cleanup

noamnelke commented on p2p/net/msgcon.go:
```diff
@@ -36,19 +37,25 @@ type MsgConnection struct {
+		remoteAddr: conn.RemoteAddr(),
+		r:          conn,
+		w:          conn,
+		//closer:       conn,
```
cleanup

noamnelke commented on p2p/net/msgcon.go:
```diff
@@ -192,15 +197,17 @@ func (c *MsgConnection) closeUnlocked() error {
 		return ErrAlreadyClosed
 	}
 	c.closed = true
+	//if err := c.closer.Close(); err != nil {
+	//	c.logger.With().Error("error closing connection io", log.Err(err))
+	//}
```
cleanup

lrettig committed:
Return conn token if incoming conn fails

Spotter: @y0sher

lrettig committed:
Formatting

lrettig committed:
Add regression test for net max pending conn

lrettig committed:
Fix concurrency issue in test

lrettig committed:
Linter

lrettig committed:
Minor test cleanup

lrettig committed:
Add shutdowns to tests

y0sher commented on p2p/net/network.go:
```diff
@@ -303,20 +307,29 @@ func (n *Net) accept(listen net.Listener) {
+	}
+
+	// The connection is automatically closed when there's no more to read, no need to close it here
+	conn.beginEventProcessing()
```
isn't this a blocking function so the `defer` call above will never be executed until `beginEventProcessing` terminates which is basically when the connection is closed ?

lrettig replied:
yes, it's blocking, that's why this whole function gets run inside a goroutine. Strictly speaking we don't need a defer here but I still think it's cleaner, more idiomatic go: this is a form of cleanup that should happen even if, e.g., there's a panic in `conn.beginEventProcessing()`. Let me know if you disagree.

y0sher replied:
the defer works and idiomatic go because we have 2 return cases. though the actual problem I see is that we're not releasing `pending` channel until the connection is closed, this was not what this code was intended to do once the connection is running and open it is not pending anymore and we don't need to hold the channel.

y0sher replied:
I agree about cleaning up and the panic case though.

lrettig replied:
Not sure I understand. The reason this is a memory leak is that we were releasing `pending` before closing the connection. If we do this, there can be any number of such connections at one time. Look here, you can see there are hundreds of them: https://github.com/spacemeshos/go-spacemesh/issues/2313#issuecomment-808552000. This caused our managed miners to run out of memory and crash, as far as I can tell. That's why I modified this not to release until the connection closes. The current testnet is running with this code and seems stable!

lrettig committed:
Merge branch 'develop' into p2p-fixleaks

lrettig commented:
bors merge
